packages/splitter/src/useSplitter.ts

```diff
@@ -0,0 +1,274 @@
+/**
+ * Copyright Zendesk, Inc.
+ *
+ * Use of this source code is governed under the Apache License, Version 2.0
+ * found at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+import { composeEventHandlers } from '@zendeskgarden/container-utilities';
+import React, { HTMLProps, MouseEvent, useCallback, useEffect, useRef, useState } from 'react';
+
+export enum SplitterOrientation {
+  HORIZONTAL = 'horizontal',
+  VERTICAL = 'vertical'
+}
+
+export enum SplitterType {
+  FIXED = 'fixed',
+  VARIABLE = 'variable'
+}
+export interface IUseSplitterProps extends Omit<HTMLProps<any>, 'onChange' | 'controls'> {
+  /** A label for the separator */
+  label?: string;
+  /** An id identifying the element labeling the separator */
+  labelledBy?: string;
+  /** An id of the element the separator controls */
+  controls: string;
+  /** A document environment to attach events to */
+  environment?: Pick<Document, 'addEventListener' | 'removeEventListener'>;
+  /** A default value for starting uncontrolled separator location */
+  defaultValueNow?: number;
+  /** Determines whether a separator behaves in fixed or variable mode */
+  type: SplitterType;
+  /** A minimum number for separator location */
+  min: number;
+  /** A maximum number for separator location */
+  max: number;
+  /** Determines whether the separator is vertical or horizontal orientation */
+  orientation: SplitterOrientation;
+  /** Determines current position of a controlled separator */
+  valueNow?: number;
+  /** A callback that receives updated valueNow when a controlled separator moves */
+  onChange?: (valueNow: number) => void;
+}
+
+export interface IUseSplitterReturnValue {
+  getSeparatorProps: <T>(options?: T & HTMLProps<any>) => any;
+}
+
+export interface IUsePrecachedStateProps<T> {
+  /* A default value for uncontrolled mode starting value */
+  defaultValue: T;
+  /* A current value for controlled state mode */
+  value?: T;
+  /* A callback that recieves updated value when controlled state changes */
+  onChange?: (value: T) => void;
+}
+
+// Problem: Generic DOM events fire independently from React state lifecycle.
+// Sometimes this will cause React state changes driven by global events to lag behind the current DOM value
+// usePrecachedState will provide "safe" state handling using a pre-cached Ref when listening to global DOM events to keep up with latest changes.
+// Eventually React State will catch up and the virtual DOM will be consistent with the DOM when using this state hook.
+// TODO Promote to Container Utilities?
+export const usePrecachedState = <T>({
```

Comment:
A new utility for Container Utilities perhaps?

Reply:
Yeah, I think that would be great. Please note, that would need to happen under a separate PR and go through a npm publish cycle so that the package-imported hook can be reused here.

Reply:
Noice! I will start working on simplifying things down since we don't need some of that extra cruft of `getBoundingClientRect`.